Rewrite `test-float-parse` in Rust

tgross35 opened the pull request:
Migrate from the currently broken Rust + Python `test-float-parse` to a Rust implementation. This newer version should be significantly faster (tests execute in parallel with threads, rather than series across multiple processes, which also eliminates the "...the worker processes are leaked and stick around forever" message), and should be significantly easier to extend to the new float types.

Since this is faster and hopefully more stable, we should be able to launch it with `x` and run the faster tests in CI.

<!-- homu-ignore:start -->

(please don't merge before I can squash this)

<!-- homu-ignore:end -->

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/9854278814/job/27206694209) [(plain)](https://github.com/rust-lang/rust/commit/74bfcda6377ecf2fff9c50495907819afada32a3/checks/27206694209/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Getting action download info
Download action repository 'msys2/setup-msys2@v2.22.0' (SHA:cc11e9188b693c2b100158c3322424c4cc1dadea)
Download action repository 'actions/checkout@v4' (SHA:692973e3d937129bcbf40652eb9f2f61becf3332)
Download action repository 'actions/upload-artifact@v4' (SHA:0b2256b8c012f0828dc542b3febcab082c67f72b)
Complete job name: PR - mingw-check-tidy
git config --global core.autocrlf false
shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
---
COPY scripts/sccache.sh /scripts/
RUN sh /scripts/sccache.sh

COPY host-x86_64/mingw-check/reuse-requirements.txt /tmp/
RUN pip3 install --no-deps --no-cache-dir --require-hashes -r /tmp/reuse-requirements.txt \
    && pip3 install virtualenv
COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---

#12 [5/8] COPY host-x86_64/mingw-check/reuse-requirements.txt /tmp/
#12 DONE 0.0s

#13 [6/8] RUN pip3 install --no-deps --no-cache-dir --require-hashes -r /tmp/reuse-requirements.txt     && pip3 install virtualenv
#13 0.408   Downloading binaryornot-0.4.4-py2.py3-none-any.whl (9.0 kB)
#13 0.419 Collecting boolean-py==4.0
#13 0.422   Downloading boolean.py-4.0-py3-none-any.whl (25 kB)
#13 0.434 Collecting chardet==5.1.0
---
#13 3.416 Building wheels for collected packages: reuse
#13 3.417   Building wheel for reuse (pyproject.toml): started
#13 3.740   Building wheel for reuse (pyproject.toml): finished with status 'done'
#13 3.741   Created wheel for reuse: filename=reuse-1.1.0-cp310-cp310-manylinux_2_35_x86_64.whl size=181117 sha256=f5f58750481f69515c2c0d1d503daf565e2565c370d07fc6aeb95fe3498b4269
#13 3.741   Stored in directory: /tmp/pip-ephem-wheel-cache-yeai5js_/wheels/c2/3c/b9/1120c2ab4bd82694f7e6f0537dc5b9a085c13e2c69a8d0c76d
#13 3.744 Installing collected packages: boolean-py, binaryornot, setuptools, reuse, python-debian, markupsafe, license-expression, jinja2, chardet
#13 3.767   Attempting uninstall: setuptools
#13 3.768     Found existing installation: setuptools 59.6.0
#13 3.769     Not uninstalling setuptools at /usr/lib/python3/dist-packages, outside environment /usr
#13 3.769     Not uninstalling setuptools at /usr/lib/python3/dist-packages, outside environment /usr
#13 3.769     Can't uninstall 'setuptools'. No files were found to uninstall.
#13 4.439 Successfully installed binaryornot-0.4.4 boolean-py-4.0 chardet-5.1.0 jinja2-3.1.2 license-expression-30.0.0 markupsafe-2.1.1 python-debian-0.1.49 reuse-1.1.0 setuptools-66.0.0
#13 4.439 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#13 4.959 Collecting virtualenv
#13 4.996   Downloading virtualenv-20.26.3-py3-none-any.whl (5.7 MB)
#13 5.062      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 5.7/5.7 MB 89.4 MB/s eta 0:00:00
#13 5.113 Collecting platformdirs<5,>=3.9.1
#13 5.116   Downloading platformdirs-4.2.2-py3-none-any.whl (18 kB)
#13 5.149 Collecting filelock<4,>=3.12.2
#13 5.152   Downloading filelock-3.15.4-py3-none-any.whl (16 kB)
#13 5.168 Collecting distlib<1,>=0.3.7
#13 5.171   Downloading distlib-0.3.8-py2.py3-none-any.whl (468 kB)
#13 5.178      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 468.9/468.9 KB 102.3 MB/s eta 0:00:00
#13 5.266 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#13 5.440 Successfully installed distlib-0.3.8 filelock-3.15.4 platformdirs-4.2.2 virtualenv-20.26.3
#13 DONE 5.5s

#14 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#14 DONE 0.0s
---
DirectMap4k:      198592 kB
DirectMap2M:     8189952 kB
DirectMap1G:    10485760 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
    Finished `dev` profile [unoptimized] target(s) in 0.03s
##[endgroup]
downloading https://ci-artifacts.rust-lang.org/rustc-builds-alt/5be2ec7245de6e8015a0daedf24d20e3e5bf2062/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz
extracting /checkout/obj/build/cache/llvm-5be2ec7245de6e8015a0daedf24d20e3e5bf2062-true/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz to /checkout/obj/build/x86_64-unknown-linux-gnu/ci-llvm
---
fmt check
fmt: checked 5240 files
tidy check
tidy: Skipping binary file check, read-only filesystem
tidy error: `/checkout/src/etc/test-float-parse/src/validate.rs:338` contains `#[test]`; unit tests and benchmarks must be placed into separate files or directories named `tests.rs`, `benches.rs`, `tests` or `benches`
##[error]tidy error: /checkout/src/bootstrap/src/core/build_steps/test.rs:3523: TODO is used for tasks that should be done before merging a PR; If you want to leave a message in the codebase use FIXME
##[error]tidy error: /checkout/src/etc/test-float-parse/README.md:5: TODO is used for tasks that should be done before merging a PR; If you want to leave a message in the codebase use FIXME
removing old virtual environment
creating virtual environment at '/checkout/obj/build/venv' using 'python3.10'
Requirement already satisfied: pip in ./build/venv/lib/python3.10/site-packages (24.1)
  Downloading pip-24.1.2-py3-none-any.whl.metadata (3.6 kB)
Downloading pip-24.1.2-py3-none-any.whl (1.8 MB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.8/1.8 MB 82.7 MB/s eta 0:00:00
Installing collected packages: pip
---
checking C++ file formatting
some tidy checks failed


Command BootstrapCommand { command: LD_LIBRARY_PATH="/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-bootstrap-tools/x86_64-unknown-linux-gnu/release/deps:/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/lib" RUSTC="/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/rustc" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/rust-tidy" "/checkout" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "/checkout/obj/build" "4" "--extra-checks=py:lint,cpp:fmt", failure_behavior: DelayFail, stdout: Print, stderr: Print, run_always: false } did not execute successfully.

Build completed unsuccessfully in 0:01:11
  local time: Tue Jul  9 09:17:12 UTC 2024
  network time: Tue, 09 Jul 2024 09:17:12 GMT
```

</details>

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/9862475546/job/27233146383) [(plain)](https://github.com/rust-lang/rust/commit/5fe2286321f07a8ea7ef124c42b960dff0500625/checks/27233146383/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Getting action download info
Download action repository 'msys2/setup-msys2@v2.22.0' (SHA:cc11e9188b693c2b100158c3322424c4cc1dadea)
Download action repository 'actions/checkout@v4' (SHA:692973e3d937129bcbf40652eb9f2f61becf3332)
Download action repository 'actions/upload-artifact@v4' (SHA:0b2256b8c012f0828dc542b3febcab082c67f72b)
Complete job name: PR - mingw-check-tidy
git config --global core.autocrlf false
shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
---
COPY scripts/sccache.sh /scripts/
RUN sh /scripts/sccache.sh

COPY host-x86_64/mingw-check/reuse-requirements.txt /tmp/
RUN pip3 install --no-deps --no-cache-dir --require-hashes -r /tmp/reuse-requirements.txt \
    && pip3 install virtualenv
COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---

#12 [5/8] COPY host-x86_64/mingw-check/reuse-requirements.txt /tmp/
#12 DONE 0.0s

#13 [6/8] RUN pip3 install --no-deps --no-cache-dir --require-hashes -r /tmp/reuse-requirements.txt     && pip3 install virtualenv
#13 0.644   Downloading binaryornot-0.4.4-py2.py3-none-any.whl (9.0 kB)
#13 0.661 Collecting boolean-py==4.0
#13 0.667   Downloading boolean.py-4.0-py3-none-any.whl (25 kB)
#13 0.684 Collecting chardet==5.1.0
---
#13 3.807 Building wheels for collected packages: reuse
#13 3.808   Building wheel for reuse (pyproject.toml): started
#13 4.128   Building wheel for reuse (pyproject.toml): finished with status 'done'
#13 4.129   Created wheel for reuse: filename=reuse-1.1.0-cp310-cp310-manylinux_2_35_x86_64.whl size=181117 sha256=f5f58750481f69515c2c0d1d503daf565e2565c370d07fc6aeb95fe3498b4269
#13 4.129   Stored in directory: /tmp/pip-ephem-wheel-cache-vcvrcsqw/wheels/c2/3c/b9/1120c2ab4bd82694f7e6f0537dc5b9a085c13e2c69a8d0c76d
#13 4.132 Installing collected packages: boolean-py, binaryornot, setuptools, reuse, python-debian, markupsafe, license-expression, jinja2, chardet
#13 4.154   Attempting uninstall: setuptools
#13 4.154     Found existing installation: setuptools 59.6.0
#13 4.156     Not uninstalling setuptools at /usr/lib/python3/dist-packages, outside environment /usr
---
#13 5.379   Downloading virtualenv-20.26.3-py3-none-any.whl (5.7 MB)
#13 5.580      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 5.7/5.7 MB 28.7 MB/s eta 0:00:00
#13 5.639 Collecting filelock<4,>=3.12.2
#13 5.646   Downloading filelock-3.15.4-py3-none-any.whl (16 kB)
#13 5.678 Collecting platformdirs<5,>=3.9.1
#13 5.684   Downloading platformdirs-4.2.2-py3-none-any.whl (18 kB)
#13 5.709 Collecting distlib<1,>=0.3.7
#13 5.716   Downloading distlib-0.3.8-py2.py3-none-any.whl (468 kB)
#13 5.725      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 468.9/468.9 KB 60.6 MB/s eta 0:00:00
#13 5.815 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#13 5.984 Successfully installed distlib-0.3.8 filelock-3.15.4 platformdirs-4.2.2 virtualenv-20.26.3
#13 DONE 6.1s

#14 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#14 DONE 0.0s
---
DirectMap4k:      190400 kB
DirectMap2M:     8198144 kB
DirectMap1G:    10485760 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
    Finished `dev` profile [unoptimized] target(s) in 0.03s
##[endgroup]
downloading https://ci-artifacts.rust-lang.org/rustc-builds-alt/f25e92bd42b14e45440c0a30a4ed751ea502f430/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz
extracting /checkout/obj/build/cache/llvm-f25e92bd42b14e45440c0a30a4ed751ea502f430-true/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz to /checkout/obj/build/x86_64-unknown-linux-gnu/ci-llvm
---
fmt check
fmt: checked 5241 files
tidy check
tidy: Skipping binary file check, read-only filesystem
tidy error: `/checkout/src/etc/test-float-parse/src/validate.rs:338` contains `#[test]`; unit tests and benchmarks must be placed into separate files or directories named `tests.rs`, `benches.rs`, `tests` or `benches`
##[error]tidy error: /checkout/src/bootstrap/src/core/build_steps/test.rs:3523: TODO is used for tasks that should be done before merging a PR; If you want to leave a message in the codebase use FIXME
##[error]tidy error: /checkout/src/etc/test-float-parse/README.md:20: TODO is used for tasks that should be done before merging a PR; If you want to leave a message in the codebase use FIXME
removing old virtual environment
creating virtual environment at '/checkout/obj/build/venv' using 'python3.10'
Requirement already satisfied: pip in ./build/venv/lib/python3.10/site-packages (24.1)
  Downloading pip-24.1.2-py3-none-any.whl.metadata (3.6 kB)
Downloading pip-24.1.2-py3-none-any.whl (1.8 MB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.8/1.8 MB 52.6 MB/s eta 0:00:00
Installing collected packages: pip
---
checking C++ file formatting
some tidy checks failed


Command BootstrapCommand { command: LD_LIBRARY_PATH="/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-bootstrap-tools/x86_64-unknown-linux-gnu/release/deps:/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/lib" RUSTC="/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/rustc" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/rust-tidy" "/checkout" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "/checkout/obj/build" "4" "--extra-checks=py:lint,cpp:fmt", failure_behavior: DelayFail, stdout: Print, stderr: Print, run_always: false } did not execute successfully.

Build completed unsuccessfully in 0:01:12
  local time: Tue Jul  9 18:47:22 UTC 2024
  network time: Tue, 09 Jul 2024 18:47:22 GMT
```

</details>

tgross35 commented:
@rustbot label +rla-silenced

Mark-Simulacrum commented on src/bootstrap/src/core/build_steps/test.rs:
```diff
@@ -3466,3 +3466,86 @@ impl Step for CodegenGCC {
+        );
+
+        // Run the actual parse tests.
+        // TODO: can we ensure this always runs with `--release`? Running without it is way
```
I don't think so, but in general optimizations are basically always on for everything compiled by bootstrap (https://github.com/rust-lang/rust/blob/366e558c3be8936bdbf3f733cb5eb48a4feb2ae4/src/bootstrap/src/core/builder.rs#L1405-L1411) and I'd not expect that to change soon. I don't think a TODO here is useful, the same can be said for basically all of our tools (e.g., the compiler).

tgross35 replied:
Okay, great. I have dependencies as opt-level=3 anyway even for debug builds, which helps. 

Yeah the TODO was just to make sure it came up in review, removed.

Mark-Simulacrum commented on src/etc/test-float-parse/Cargo.lock:
```diff
@@ -2,28 +2,195 @@
 # It is not intended for manual editing.
 version = 3
 
+[[package]]
```
Generally speaking tools are members of the global workspace, you probably want to add this one to the list: https://github.com/rust-lang/rust/blob/master/Cargo.toml#L3

That helps ensure that e.g. license checking is covered for their dependencies and they don't need special exceptions (e.g., https://github.com/rust-lang/rust/blob/fcaa6fdfbee1316184e7ad98c53241d52cd30a5f/src/tools/tidy/src/deps.rs#L68)

tgross35 replied:
Think this should be done.

Mark-Simulacrum commented on src/etc/test-float-parse/Cargo.toml:
```diff
@@ -8,7 +8,24 @@ publish = false
 resolver = "1"
```
Why is this using an old resolver?

tgross35 replied:
I just didn't update what was already there. This doesn't even need to be a workspace anymore, I just removed the block.

Mark-Simulacrum commented on src/etc/test-float-parse/Cargo.toml:
```diff
@@ -8,7 +8,24 @@ publish = false
+
+[profile.profiling]
+inherits = "release"
+debug = true
```
I guess the thinking is that the super long runtime on these tests makes this merited above the normal status quo? I'd rather see these move to bootstrap (e.g., by setting Cargo env variables for configuring these properties) or into the root Cargo.toml per a previous comment, both of which I believe would retain the customization that does seem justified for multi-hour test runs.

If those options are too annoying/difficult, then let's forgo moving this crate into the workspace and instead make sure it's part of the tidy dependency checking.

tgross35 replied:
Ah I only had this for when I was trying to debug a slow test. I’ll remove it.

Mark-Simulacrum commented on src/etc/test-float-parse/src/gen/exhaustive.rs:
```diff
@@ -0,0 +1,42 @@
+
+use crate::{Float, Generator, Int};
+
+/// Test every possible bit pattern. Not recommended for anything larger than `f32`!
```
Is "not recommended" really the right phrasing? It's ~impossible for a 2^64 test to complete, and we don't have intermediate types like ~f48 or something.

tgross35 replied:
Good point, I updated the wording here

Mark-Simulacrum commented on src/bootstrap/src/core/build_steps/test.rs:
```diff
@@ -3466,3 +3466,86 @@ impl Step for CodegenGCC {
+        let cargo_test = tool::prepare_tool_cargo(
+            builder,
+            compiler,
+            // todo: seems like this should be `Mode::Std`
```
Actually, maybe it is different for this tool - are we wanting to test any functionality implemented in std? I guess the answer is yes, in which case this should definitely be Mode::Std.

tgross35 replied:
What is the distinction between the two? This only needs to test functionality in the standard library so building the compiler shouldn’t be required. When I tried setting Mode::Std it said it couldn’t find the standard library (recommended build-std), so I must be missing something.

Mark-Simulacrum replied:
Sorry, you probably want `Mode::ToolStd`. See documentation in https://github.com/rust-lang/rust/blob/73a228116ae8c8ce73e309eee8c730ce90feac78/src/bootstrap/src/lib.rs#L222-L233

ToolBootstrap uses a beta std, ToolStd uses a local std.

tgross35 commented:
cc @Alexhuszagh by the way, since you are the original author

tgross35 commented on src/bootstrap/src/core/build_steps/test.rs:
```diff
@@ -3466,3 +3466,79 @@ impl Step for CodegenGCC {
+            &[],
+        );
+
+        run_cargo_test(
+            cargo_test,
+            &[],
+            &[],
+            crate_name,
+            crate_name,
+            compiler,
+            bootstrap_host,
+            builder,
+        );
+
+        // Run the actual parse tests.
+        let mut cargo_run = tool::prepare_tool_cargo(
+            builder,
+            compiler,
+            Mode::ToolBootstrap,
+            bootstrap_host,
+            "run",
+            path,
+            SourceType::InTree,
+            &[],
+        );
+
+        cargo_run.arg("--");
+        if builder.config.args().is_empty() {
+            // By default, exclude tests that take longer than ~1m.
+            cargo_run.arg("--skip-huge");
+        } else {
+            cargo_run.args(builder.config.args());
+        }
+
+        cargo_run.into_cmd().run(builder);
```
Because `run_cargo_test` automatically passes the builder args to `cargo test`, there is a bit of an annoyance here that `./x t src/etc/test-float-parse -- somearg` can't pass arbitrary arguments to the `test-float-parse` binary since they get rejected by `cargo test`. Not a huge problem since you can run it directly from the directory, but do we have anything else that does something with arguments like this @Mark-Simulacrum?

Other than that I think I have finally tapered my edits here, pretty happy with the current state.

Mark-Simulacrum replied:
x.py run may make more sense if you want to run a specific tool.

tgross35 replied:
Didn't realize that worked too, great. Guess this is all set then.

Mark-Simulacrum commented:
r=me with ToolStd (presuming that works) so we're actually testing the in-tree std.

tgross35 commented:
Setting to `ToolStd` says `` can't find crate for `core` `` and `std`, at least locally for me (waiting to see what CI says). Is there something specific needed to make this work?

I also needed to move the optimization config to workspace `Cargo.toml` to avoid a warning.

tgross35 committed:
Rewrite `test-float-parse` in Rust

The existing implementation uses Python to launch a set of Rust-written
binaries. Unfortunately, this is currently broken; it seems that some
updates meant it no longer compiles.

There is also a problem that support for more float types (`f16`,
`f128`) would be difficult to add since this is very specialized to
`f32` and `f64`.

Because of these sortcomings, migrate to a version written in Rust. This
version should be significantly faster; test generators can execute in
parallel, and test cases are chunked and parallelized. This should also
resolve the preexisting "... the worker processes are leaked and stick
around forever" comment.

This change also introduces genericism over float types and properties,
meaning it will be much easier to extend support to newly added types.

`num::BigRational` is used in place of Python's fractions for
infinite-precision calculations.

tgross35 committed:
Move `test-float-parse` to the global workspace

Since `test-float-parse` is now implemented in Rust, we can move it into
the global workspace and check dependency licenses.

tgross35 committed:
Expose `test-float-parse` via bootstrap

With updates to `test-float-parse`, it is now possible to run as another
Rust tool. Enable check, clippy, and test.

Test runs the unit tests, as well as shorter parsing tests (takes
approximately 1 minute).

tgross35 committed:
Run `test-float-parse` as part of CI

With the previous improvements, it is now possible to run float parsing
tests as part of CI. Enable it here.

This only runs a subset of tests, which takes about one minute.

Mark-Simulacrum commented:
```
$ git diff
diff --git a/src/bootstrap/src/core/build_steps/test.rs b/src/bootstrap/src/core/build_steps/test.rs
index 96472587476..9692d502622 100644
--- a/src/bootstrap/src/core/build_steps/test.rs
+++ b/src/bootstrap/src/core/build_steps/test.rs
@@ -3494,6 +3494,8 @@ fn run(self, builder: &Builder<'_>) {
         let path = self.path.to_str().unwrap();
         let crate_name = self.path.components().last().unwrap().as_os_str().to_str().unwrap();

+        builder.ensure(compile::Std::new(compiler, bootstrap_host));
+
         // Run any unit tests in the crate
         let cargo_test = tool::prepare_tool_cargo(
             builder,
```

Fixes it for me... not sure exactly why that's not automatic but I think this should fix it.

tgross35 commented:
I just came to the same conclusion, thanks, great that it works now. I fixed this then rebased since I know there have been some bootstrap updates recently. I'll r=you after CI passes unless you have other concerns.

tgross35 commented:
Thanks for reviewing all of this!

@bors r=Mark-Simulacrum

bors commented:
:pushpin: Commit 7f7ec2d438b831f0ed893c0a9720ad6c2db38adb has been approved by `Mark-Simulacrum`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=Mark-Simulacrum 7f7ec2d438b831f0ed893c0a9720ad6c2db38adb -->
<!-- homu: {"type":"Approved","sha":"7f7ec2d438b831f0ed893c0a9720ad6c2db38adb","approver":"Mark-Simulacrum","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

tgross35 commented:
Also opened an issue to adjust bootstrap behavior at some point https://github.com/rust-lang/rust/issues/128012.

tgross35 commented:
@bors r-
Failed the rollup https://github.com/rust-lang/rust/pull/128024#issuecomment-2241482396

tgross35 commented:
I made a typo in the path to test, `etc/etc/test-float-parse` rather than `src/etc/test-float-parse`. Just a trivial fix so

@bors r=Mark-Simulacrum

bors commented:
:pushpin: Commit d12387835f32ee333bc41b85a43234f39f7e32ef has been approved by `Mark-Simulacrum`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=Mark-Simulacrum d12387835f32ee333bc41b85a43234f39f7e32ef -->
<!-- homu: {"type":"Approved","sha":"d12387835f32ee333bc41b85a43234f39f7e32ef","approver":"Mark-Simulacrum","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
